homeassistant/components/fritzbox/__init__.py

```diff
@@ -11,80 +11,89 @@
-                                vol.Required(CONF_USERNAME): cv.string,
-                            }
-                        )
-                    ],
+        DOMAIN: vol.All(
+            cv.ensure_list,
+            [
+                vol.Schema(
+                    {
+                        vol.Required(CONF_HOST, default=DEFAULT_HOST): cv.string,
+                        vol.Required(CONF_PASSWORD): cv.string,
+                        vol.Required(
+                            CONF_USERNAME, default=DEFAULT_USERNAME
+                        ): cv.string,
+                    }
                 )
-            }
+            ],
+            ensure_unique_hosts,
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
 
-def setup(hass, config):
-    """Set up the fritzbox component."""
+async def async_setup(hass, config):
+    """Set up the AVM Fritz!Box integration."""
+    if DOMAIN in config:
+        for entry_config in config[DOMAIN]:
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN, context={"source": "import"}, data=entry_config
+                )
+            )
+
+    return True
+
 
-    fritz_list = []
+async def async_setup_entry(hass, entry):
+    """Set up the AVM Fritz!Box platforms."""
+    fritz = Fritzhome(
+        host=entry.data[CONF_HOST],
+        user=entry.data[CONF_USERNAME],
+        password=entry.data[CONF_PASSWORD],
+    )
+    await hass.async_add_executor_job(fritz.login)
 
-    configured_devices = config[DOMAIN].get(CONF_DEVICES)
-    for device in configured_devices:
-        host = device.get(CONF_HOST)
-        username = device.get(CONF_USERNAME)
-        password = device.get(CONF_PASSWORD)
-        fritzbox = Fritzhome(host=host, user=username, password=password)
-        try:
-            fritzbox.login()
-            _LOGGER.info("Connected to device %s", device)
-        except LoginError:
-            _LOGGER.warning("Login to Fritz!Box %s as %s failed", host, username)
-            continue
+    hass.data.setdefault(DOMAIN, {CONF_CONNECTIONS: {}, CONF_DEVICES: set()})
+    hass.data[DOMAIN][CONF_CONNECTIONS][entry.entry_id] = fritz
 
-        fritz_list.append(fritzbox)
+    for domain in SUPPORTED_DOMAINS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, domain)
+        )
 
-    if not fritz_list:
-        _LOGGER.info("No fritzboxes configured")
-        return False
+    def logout_fritzbox(event):
+        """Close connections to this fritzbox."""
+        fritz.logout()
 
-    hass.data[DOMAIN] = fritz_list
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, logout_fritzbox)
 
-    def logout_fritzboxes(event):
-        """Close all connections to the fritzboxes."""
-        for fritz in fritz_list:
-            fritz.logout()
+    return True
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, logout_fritzboxes)
```

Comment:
Is calling `fritz.logout` here directly also possible?

Reply:
How do you mean?

Reply:
Instead of using `logout_fritzbox`, so we can avoid this helper function.

Reply:
I think we need to keep it since the callback needs to accept the event time argument.